src/DFMM.sol

```diff
@@ -108,24 +110,21 @@ contract DFMM is IDFMM {
 
         for (uint256 i = 0; i < tokensLength; i++) {
             address token = params.tokens[i];
-
-            for (uint256 j = 0; j < tokensLength; j++) {
-                if (i != j && token == params.tokens[j]) {
-                    revert InvalidDuplicateTokens();
-                }
-            }
-        }
-
-        for (uint256 i = 0; i < tokensLength; i++) {
             uint256 decimals = ERC20(params.tokens[i]).decimals();
 
             if (decimals > 18 || decimals < 6) {
                 revert InvalidTokenDecimals();
             }
 
-            _transferFrom(params.tokens[i], reserves[i]);
+            for (uint256 j = i + 1; j < tokensLength; j++) {
```

Comment:
I'll note that you could also wrap with `unchecked` here as otherwise the `i + 1` operation will be checked for overflow (unnecessarily, given the limit placed on `tokens.length` above). Depending on the exact Solidity version used, the loop increments may or may not be checked by default as well (in `0.8.22` they introduced unchecked increments as a feature).